File: WEEK5_23_2_26/ErrorHandling.java
public class ErrorHandling {
    // Try to parse inputs and divide; return descriptive errors on failure
    public static String safeDivide(String aStr, String bStr) {
        try {
            double a = Double.parseDouble(aStr);
            double b = Double.parseDouble(bStr);
            if (b == 0.0) return "Error: division by zero";
            return String.valueOf(a / b);
        } catch (NumberFormatException e) {
            return "Error: unsupported operand type";
        }
    }

    // Parse integer or throw a clear exception
    public static int readInt(String value) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException e) {
            throw new NumberFormatException("Invalid integer: " + value);
        }
    }

    public static void main(String[] args) {
        // Demo calls
        System.out.println("safeDivide(10,2) -> " + safeDivide("10", "2"));
        System.out.println("safeDivide(5,0) -> " + safeDivide("5", "0"));
        System.out.println("safeDivide(\"a\",2) -> " + safeDivide("a", "2"));

        String[] vals = {"3", "x", "5"};
        for (String v : vals) {
            try {
                System.out.println("readInt(" + v + ") -> " + readInt(v));
            } catch (NumberFormatException e) {
                System.out.println("readInt error: " + e.getMessage());
            }
        }
    }
}

File: WEEK5_23_2_26/error_handling.py
"""Simple error handling examples.
Keep functions small and demonstrative.
"""
# Return quotient or an explanatory error string when invalid.
def safe_divide(a, b):
    try:
        return a / b
    except ZeroDivisionError:
        return "Error: division by zero"
    except TypeError:
        return "Error: unsupported operand type"


# Convert value to int or raise a clear ValueError.
def read_int(value):
    try:
        return int(value)
    except ValueError:
        raise ValueError("Invalid integer: {}".format(value))


def main():
    # Demo calls showing normal and error cases
    print('safe_divide(10, 2) ->', safe_divide(10, 2))
    print('safe_divide(5, 0) ->', safe_divide(5, 0))
    print('safe_divide("a", 2) ->', safe_divide('a', 2))

    for val in ['3', 'x', 5.6]:
        try:
            print('read_int({}) ->'.format(val), read_int(val))
        except ValueError as e:
            print('read_int error:', e)


if __name__ == '__main__':
    main()
